### PR TITLE
Add possibility to register a regexp in router.

### PR DESCRIPTION
**Pull request by klen**
I've added possibility to register any regexp in `aiohttp.web.UrlDispatcher`. For that moment there is no opportunity to create some dynamic routes like `/test/(?P<any_include_slashes>*.)` or maybe something like this `/test/(.*)/?`. The little request implement the functionality.

    # common aiohttp usage
    app.router.add_route('GET', '/some', handler)

    import re

    # Define regexp manually
    app.router.add_route('GET', re.compile('^/any/regexp/here/?$'), handler)

### CODE REVIEW

**Comment by asvetlov** [comment]
-1

You already can use construction like:

```
self.router.add_route('GET', r'/handler/{to:.+}/tail', handler)
```

**Comment by klen** [comment]
No, I can't.

What should I do if I need regexp like this by example:

`/resource/([^/]+)?/?`

Looks like current solution is:

```
 handler = ...
 app.router.add_route('GET', '/resource', handler)
 app.router.add_route('GET', '/resource/', handler)
 app.router.add_route('GET', '/resource/{resource}', handler)
 app.router.add_route('GET', '/resource/{resource}/', handler)
```

Of course for current version users could do that:

```
from aiohttp.web import DynamicRoute
import re

pattern = re.compile('...')
app.router._register_endpoint(DynamicRoute('GET', pattern.pattern, handler, pattern))
```

But it's little bit ugly. Much better `app.router.register('GET', re.compile('...'))` isnt?

What the problem to provide for  users more control, which breaks nothing?

**Comment by asvetlov** [comment]
At least it breaks reverse url building by `route.url()` call.
Regexp _is not_ a formatter string.

**Comment by asvetlov** [comment]
Also, what the difference between `/resource/([^/]+)?/?` and `r'/handler/{tail:([^/]+)?/?}`

**Comment by klen** [comment]
> At least it breaks reverse url building by route.url() call.

I have not thought about `route.url`, it could be problem coz. But I see some solutions for this.

> Also, what the difference between /resource/([^/]+)?/? and r'/handler/{tail:([^/]+)?/?}

Second pattern doesn't work in current version aiohttp (because https://github.com/KeepSafe/aiohttp/blob/master/aiohttp/web.py#L1448). So that the main reason why I've made the request for support custom regexp.

**Comment by asvetlov** [comment]
Aha, I see.
I'm strongly -1 for pushing regexp instances into `.add_route()` but ok to relax route pattern rules to pass regex with slashes (IIRC Pyramid allows url patterns like this).

**Comment by klen** [comment]
And looks like route.url is already broken:

```
from aiohttp.web import UrlDispatcher as Router

route = Router().add_route('GET', r'/{one}/{two:.+}', lambda r: None)

# will raise ValueError
route.url({'one': 1, 'two': 2})
```

**Comment by asvetlov** [comment]
Sure, it works:

```
def test_route_dynamic_with_regex(self):
    handler = self.make_handler()
    route = self.router.add_route('GET', r'/{one}/{two:.+}', handler)

    url = route.url(parts={'one': 1, 'two': 2})
    self.assertEqual('/1/2', url)
```

**Comment by klen** [comment]
Oh, I have look to version in pypi before #264

**Comment by klen** [comment]
Ok, please have a look on the last solution. It makes the deal with slashes as well and pass all tests except this one:

```
def test_add_url_invalid5(self):
    handler = self.make_handler()
    with self.assertRaises(ValueError):
        self.router.add_route('post', '/post"{id}', handler)
```

Because the new dispatcher allows to define routes like this as dynamic. route.url(parts={'id': 1}) -> /post"1

File: tests/test_urldispatch.py [hunk 1]
```diff
@@ -322,9 +331,9 @@ def test_add_route_with_invalid_re(self):
         handler = self.make_handler()
         with self.assertRaises(ValueError) as ctx:
             self.router.add_route('GET', r'/handler/{to:+++}', handler)
-        s = str(ctx.exception)
-        self.assertTrue(s.startswith(
-            "Bad pattern '/handler/(?P<to>+++)': nothing to repeat"), s)
+        self.assertEqual(
+            "Bad pattern '\/handler\/(?P<to>+++)': nothing to repeat",
+            str(ctx.exception))
```
Comment (asvetlov) [inline tests/test_urldispatch.py]: Please revert the test: on Python 3.5 error message contains additional info about position in text which causes the error.

File: .gitignore [hunk 2]
```diff
@@ -1,33 +1,34 @@
-*~
```
Comment (asvetlov) [inline .gitignore:1]: Why do you reorder gitignore file?
If you need to add something -- please do it explicitly.
I just do miss your changes.

File: aiohttp/web.py [hunk 3]
```diff
@@ -1436,55 +1437,48 @@ def _register_endpoint(self, route):
...
+        formatter = ''
+        for part in self.ROUTE_RE.split(path):
             match = self.DYN.match(part)
             if match:
-                parts.append('(?P<' + match.group('var') + '>' +
-                             self.GOOD + ')')
-                factory = DynamicRoute
-                format_parts.append('{'+match.group('var')+'}')
+                pattern += '(?P<' + match.group('var') + '>' + self.GOOD + ')'
+                formatter += '{%s}' % match.group('var')
```
Comment (asvetlov) [inline aiohttp/web.py]: Please use `.format()` for string formatting: other parts of aiohttp tends to `%` style only (but we have couple exceptions now).

File: aiohttp/web.py [hunk 4]
```diff
@@ -1436,55 +1437,48 @@ def _register_endpoint(self, route):
...
                 continue
 
             match = self.DYN_WITH_RE.match(part)
             if match:
-                parts.append('(?P<' + match.group('var') + '>' +
-                             match.group('re') + ')')
-                factory = DynamicRoute
-                format_parts.append('{'+match.group('var')+'}')
+                pattern += '(?P<%(var)s>%(re)s)' % match.groupdict()
```
Comment (asvetlov) [inline aiohttp/web.py]: %

File: aiohttp/web.py [hunk 5]
```diff
@@ -1436,55 +1437,48 @@ def _register_endpoint(self, route):
...
+                formatter += '{%s}' % match.group('var')
```
Comment (asvetlov) [inline aiohttp/web.py]: %

**Comment by asvetlov** [comment]
The last patch if perfect except my tiny inline notes.

**Comment by asvetlov** [comment]
Hmm. Does you patch violate the rule "A pattern segment (an individual item between / characters in the pattern) may either be a literal string (e.g. foo) or it may be a replacement marker (e.g. {foo}) or a certain combination of both"?

**Comment by asvetlov** [comment]
Please forget about last message: I want to support patterns like `foo/{name}.html` too.

File: .gitignore (hunk 2 above)
Comment (klen) [inline .gitignore:1]: Just have added `.vimrc` and sorted in alphabetical order, because there was no any structure before.

Comment (asvetlov) [inline .gitignore:1]: Ok

**Comment by asvetlov** [comment]
Thanks!

**Pull request merged**